### PR TITLE
Create parent class for compute::kernels and compute::properties

### PR DESCRIPTION
**Pull request by icui**
# Description

Please describe the changes/features in this pull request. 

# Issue Number

If there is an issue created for these changes, link it here

# Checklist

Please make sure to check developer documentation on specfem docs. 

[] I ran the code through pre-commit to check style
[] My code passes all the integration tests
[] I have added sufficient unittests to test my changes
[] I have added/updated documentation for the changes I am proposing
[] I have updated CMakeLists to ensure my code builds
[] My code builds across all platforms

### CODE REVIEW

File: include/compute/boundary_values/impl/boundary_medium_container.tpp [hunk 1]
```diff
@@ -20,7 +20,7 @@ specfem::compute::impl::boundary_medium_container<DimensionType, MediumType,
   const int nx = mesh.ngllx;
 
   for (int ispec = 0; ispec < nspec; ispec++) {
-    if (properties.h_element_types(ispec) == MediumType &&
```
Comment (Rohit-Kakodkar) [inline include/compute/boundary_values/impl/boundary_medium_container.tpp:23]: Can you add boundary_tags to element_types? Then we can call method `get_element_indices(medium_tag, boundary_tag)` here. We also would not the dependency the properties and boundary dependency here for boundary_values.

File: include/compute/coupled_interfaces/interface_container.tpp [hunk 2]
```diff
@@ -545,10 +545,10 @@ specfem::compute::interface_container<MediumTag1, MediumTag2>::
     const int ispec1_compute = mapping.mesh_to_compute(ispec1_mesh);
     const int ispec2_compute = mapping.mesh_to_compute(ispec2_mesh);
 
-    if (!(((properties.h_element_types(ispec1_compute) == MediumTag1) &&
-           (properties.h_element_types(ispec2_compute) == MediumTag2)) ||
-          ((properties.h_element_types(ispec1_compute) == MediumTag2 &&
-            properties.h_element_types(ispec2_compute) == MediumTag1)))) {
+    if (!(((properties.h_medium_tags(ispec1_compute) == MediumTag1) &&
```
Comment (Rohit-Kakodkar) [inline include/compute/coupled_interfaces/interface_container.tpp:548]: We can again call element types here and not properties.

File: include/compute/impl/value_containers.hpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+namespace impl {
+/**
+ * @brief Values for every quadrature point in the
+ * finite element mesh
+ *
+ */
+template <template <specfem::element::medium_tag, specfem::element::property_tag>
+          class containers_type>
+struct value_containers: public element_types {
```
Comment (Rohit-Kakodkar) [inline include/compute/impl/value_containers.hpp:17]: Would be better to make properties a parent of element_types. i.e. 

```
template <template <specfem::element::medium_tag, specfem::element::property_tag>
          class containers_type>
 struct value_containers;
 
 class properties : public value_containers<...>, public element_types;
 
 class kernels : public value_containers<...>, public element_types;
 ```

File: include/compute/impl/value_containers.hpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+               const specfem::mesh::tags<specfem::dimension::type::dim2> &tags);
+  ///@}
+
+  /**
+  * @brief Returns the material_kernel for a given medium and property
+  *
+  */
+  template <specfem::element::medium_tag MediumTag,
+            specfem::element::property_tag PropertyTag>
+  const containers_type<MediumTag, PropertyTag>
```
Comment (Rohit-Kakodkar) [inline include/compute/impl/value_containers.hpp:64]: This needs to be a `KOKKOS_FUNCTION`. Should this be a constexpr?

File: include/enumerations/medium.hpp [hunk 5]
```diff
@@ -55,11 +55,11 @@ constexpr auto material_systems() {
   return material_systems;
 }
 
-constexpr auto element_types() {
-  constexpr int total_element_types = 8;
+constexpr auto medium_tags() {
+  constexpr int total_medium_tags = 8;
```
Comment (Rohit-Kakodkar) [inline include/enumerations/medium.hpp:59]: Why did you change the nomenclature here?

File: include/writer/kernel.tpp [hunk 6]
```diff
@@ -38,17 +38,17 @@ void specfem::writer::kernel<OutputLibrary>::write() {
   Kokkos::parallel_reduce(
       "specfem::writer::kernel", specfem::kokkos::HostRange(0, nspec),
       [=](const int ispec, int &n_elastic_isotropic, int &n_elastic_anisotropic, int &n_acoustic) {
-        if (kernels.h_element_types(ispec) ==
+        if (kernels.h_medium_tags(ispec) ==
```
Comment (Rohit-Kakodkar) [inline include/writer/kernel.tpp:41]: Can you call `kernels.get_element_on_host` here?

File: src/compute/compute_properties.cpp [hunk 7]
```diff
@@ -11,9 +11,9 @@ void compute_number_of_elements_per_medium(
     const int nspec, const specfem::compute::mesh_to_compute_mapping &mapping,
     const specfem::mesh::tags<specfem::dimension::type::dim2> &tags,
     const specfem::kokkos::HostView1d<specfem::element::medium_tag>
-        &h_element_types,
+        &h_medium_tags,
```
Comment (Rohit-Kakodkar) [inline src/compute/compute_properties.cpp:14]: Would be nice to create a element_types constructor where these values are generated

**Review by Rohit-Kakodkar — CHANGES_REQUESTED** [review]

File: include/enumerations/medium.hpp (hunk 5 above)
Comment (icui) [reply include/enumerations/medium.hpp]: fixed.

File: include/writer/kernel.tpp (hunk 6 above)
Comment (icui) [reply include/writer/kernel.tpp:41]: do you want to create a function count_elements_on_host()?

Comment (Rohit-Kakodkar) [reply include/writer/kernel.tpp:41]: You wont need that for this case. Since on line 73 you call `get_element_on_host` you can replace this Kokkos loop with those functions. And then use `int n_elastic_anisotropic = element_indices.extent(0)`

File: include/compute/impl/value_containers.hpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+   */
+  value_containers() = default;
+
+  /**
+   * @brief Returns the material_kernel for a given medium and property
+   *
+   */
+  template <specfem::element::medium_tag MediumTag,
+            specfem::element::property_tag PropertyTag>
+  KOKKOS_INLINE_FUNCTION const containers_type<MediumTag, PropertyTag> &
```
Comment (Rohit-Kakodkar) [inline include/compute/impl/value_containers.hpp:47]: Does it compile with `constexpr` here?

File: include/compute/properties/properties.hpp [hunk 9]
```diff
@@ -21,46 +22,10 @@ namespace compute {
...
-      specfem::element::medium_tag::elastic,
-      specfem::element::property_tag::anisotropic>
-      elastic_anisotropic; ///< Elastic anisotropic material properties
-  specfem::compute::impl::properties::material_property<
-      specfem::element::medium_tag::acoustic,
-      specfem::element::property_tag::isotropic>
-      acoustic_isotropic; ///< Acoustic isotropic material properties
-
+struct properties
+    : public impl::element_types,
```
Comment (Rohit-Kakodkar) [inline include/compute/properties/properties.hpp:26]: This is pretty nice

File: include/writer/kernel.tpp [hunk 10]
```diff
@@ -35,31 +33,14 @@ void specfem::writer::kernel<OutputLibrary>::write() {
   int n_elastic_anisotropic = 0;
```
Comment (Rohit-Kakodkar) [inline include/writer/kernel.tpp:33]: You dont need these here. Could instantiate them below

**Review by Rohit-Kakodkar — APPROVED** [review]
Minor comments. Otherwise LGTM

File: include/compute/impl/value_containers.hpp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  get_container() const {
```
Comment (lsawade) [inline include/compute/impl/value_containers.hpp:48]: This is cool

**Review by lsawade — APPROVED** [review]
LGTM
